### PR TITLE
[exporter/prometheus] fix: validate metric types and help/descriptions

### LINKED ISSUE #28617: Prometheus exporter should allow conflicting help and unit metadata
### Component(s)

_No response_

## Description

Original report: https://github.com/istio/istio/issues/47397

The prometheus exporter fails to gather metrics if the description differs between two metrics with the same name.

## Steps to Reproduce

With a prometheus receiver, scrape two endpoints with the same metric name, but different HELP (description).  Use the prometheus exporter to export the metrics.

## Expected Result

No error.  Per the [OTel specification](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#metric-metadata-1):

> Exporters MUST drop entire metrics to prevent conflicting TYPE comments, but SHOULD NOT drop metric points as a result of conflicting UNIT or HELP comments. Instead, all but one of the conflicting UNIT and HELP comments (but not metric points) SHOULD be dropped. If dropping a comment or metric points, the exporter SHOULD warn the user through error logging.

I.e. we should be accepting the conflicting description, choosing one description to keep, and one to drop, and warning the user.

## Actual Result
 
The prometheus exporter fails incoming scrapes.

### Collector version

Don't know

## Environment
OS: (e.g., "Ubuntu 20.04")
Compiler(if manually compiled): (e.g., "go 14.2")

### OpenTelemetry Collector configuration

_No response_

### Log output

```shell
2023-09-29T15:24:02.049Z        error   prometheusexporter@v0.80.0/log.go:23    error gathering metrics: 2 error(s) occurred:
* collected metric istio_build label:{name:"app"  value:"istiod"}  label:{name:"cluster"  value:"glooplatform-central"}  label:{name:"component"  value:"pilot"}  label:{name:"install_operator_istio_io_owning_resource"  value:"unknown"}  label:{name:"instance"  value:"10.108.2.7:15014"}  label:{name:"istio"  value:"istiod"}  label:{name:"istio_io_rev"  value:"1-18"}  label:{name:"job"  value:"istiod-metrics"}  label:{name:"namespace"  value:"istio-system"}  label:{name:"operator_istio_io_component"  value:"Pilot"}  label:{name:"pod_name"  value:"istiod-1-18-5f7964bccd-jlfwh"}  label:{name:"pod_template_hash"  value:"5f7964bccd"}  label:{name:"sidecar_istio_io_inject"  value:"false"}  label:{name:"tag"  value:"1.18.2-solo"}  gauge:{value:1} has help "Istio component build info" but should have ""
* collected metric istio_build label:{name:"app"  value:"istiod"}  label:{name:"cluster"  value:"glooplatform-east"}  label:{name:"component"  value:"pilot"}  label:{name:"install_operator_istio_io_owning_resource"  value:"unknown"}  label:{name:"instance"  value:"10.104.3.14:15014"}  label:{name:"istio"  value:"istiod"}  label:{name:"istio_io_rev"  value:"1-18"}  label:{name:"job"  value:"istiod-metrics"}  label:{name:"namespace"  value:"istio-system"}  label:{name:"operator_istio_io_component"  value:"Pilot"}  label:{name:"pod_name"  value:"istiod-1-18-7d66654c46-c5frb"}  label:{name:"pod_template_hash"  value:"7d66654c46"}  label:{name:"sidecar_istio_io_inject"  value:"false"}  label:{name:"tag"  value:"1.18.2-solo"}  gauge:{value:1} has help "Istio component build info" but should have ""
        {"kind": "exporter", "data_type": "metrics", "name": "prometheus"}
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusexporter.(*promLogger).Println
        github.com/open-telemetry/opentelemetry-collector-contrib/exporter/prometheusexporter@v0.80.0/log.go:23
github.com/prometheus/client_golang/prometheus/promhttp.HandlerForTransactional.func1
        github.com/prometheus/client_golang@v1.16.0/prometheus/promhttp/http.go:144
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2109
net/http.(*ServeMux).ServeHTTP
        net/http/server.go:2487
go.opentelemetry.io/collector/config/confighttp.(*decompressor).wrap.func1
        go.opentelemetry.io/collector/config/confighttp@v0.80.0/compression.go:111
net/http.HandlerFunc.ServeHTTP
        net/http/server.go:2109
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Handler).ServeHTTP
        go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.42.0/handler.go:212
go.opentelemetry.io/collector/config/confighttp.(*clientInfoHandler).ServeHTTP
        go.opentelemetry.io/collector/config/confighttp@v0.80.0/clientinfohandler.go:28
net/http.serverHandler.ServeHTTP
        net/http/server.go:2947
net/http.(*conn).serve
        net/http/server.go:1991
```
```


### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kristinapathak**
#### Description
Fixes bug where exporting fails due to different help messages for the same metric.  With this solution, the exporter will always export metrics of the same name with the first description it receives. This also rejects metrics whose types have changed. These changes follow the [spec](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#metric-metadata-1):
>Exporters MUST drop entire metrics to prevent conflicting TYPE comments, but SHOULD NOT drop metric points as a result of conflicting UNIT or HELP comments. Instead, all but one of the conflicting UNIT and HELP comments (but not metric points) SHOULD be dropped. If dropping a comment or metric points, the exporter SHOULD warn the user through error logging.

Based on https://github.com/open-telemetry/opentelemetry-go/pull/3469

#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/28617

#### Testing
Unit test cases added.

### CODE REVIEW

File: exporter/prometheusexporter/collector.go [hunk 1]
```diff
@@ -40,6 +43,7 @@ func newCollector(config *Config, logger *zap.Logger) *collector {
 		sendTimestamps:    config.SendTimestamps,
 		constLabels:       config.ConstLabels,
 		addMetricSuffixes: config.AddMetricSuffixes,
+		metricFamilies:    make(map[string]*dto.MetricFamily),
```
Comment (dashpole) [inline exporter/prometheusexporter/collector.go:46]: I think we will need to "expire" these to ensure we don't leak memory, similar to the accumulator above.

Comment (kristinapathak) [reply exporter/prometheusexporter/collector.go]: Updated to clean up expired metrics - should this also be a sync.Map?

Comment (dashpole) [reply exporter/prometheusexporter/collector.go]: Yes, it does need to be a sync.Map

**Review by dashpole — APPROVED** [review]

**Review by mwear — APPROVED** [review]

File: exporter/prometheusexporter/collector.go [hunk 2]
```diff
@@ -403,4 +428,50 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+	return emf.mf.GetHelp(), nil
+}
+
+func (c *collector) cleanupMetricFamilies() {
+	expirationTime := time.Now().Add(-c.metricExpiration)
+
+	c.metricFamilies.Range(func(key, value any) bool {
+		v := value.(metricFamily)
+		if expirationTime.After(v.lastSeen) {
+			c.logger.Debug(fmt.Sprintf("metric expired: %s", key))
```
Comment (atoulme) [inline exporter/prometheusexporter/collector.go:471]: any way to avoid calling fmt.Sprintf here

Comment (codeboten) [reply exporter/prometheusexporter/collector.go]: thanks @kristinapathak

File: exporter/prometheusexporter/collector.go [hunk 3]
```diff
@@ -403,4 +428,50 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
...
+	return emf.mf.GetHelp(), nil
+}
+
+func (c *collector) cleanupMetricFamilies() {
+	expirationTime := time.Now().Add(-c.metricExpiration)
+
+	c.metricFamilies.Range(func(key, value any) bool {
+		v := value.(metricFamily)
+		if expirationTime.After(v.lastSeen) {
+			c.logger.Debug(fmt.Sprintf("metric expired"), zap.String("instrument", key.(string)))
```
Comment (codeboten) [inline exporter/prometheusexporter/collector.go:471]:
```suggestion
			c.logger.Debug("metric expired", zap.String("instrument", key.(string)))
```

File: exporter/prometheusexporter/collector.go [hunk 4]
```diff
@@ -403,4 +428,50 @@ func (c *collector) Collect(ch chan<- prometheus.Metric) {
 		ch <- m
 		c.logger.Debug(fmt.Sprintf("metric served: %s", m.Desc().String()))
```
Comment (codeboten) [inline exporter/prometheusexporter/collector.go:429]: Doesn't need to be in this PR but as a followup would be good to clean up this sprintf as well

**Review by codeboten — APPROVED** [review]

**Pull request merged**